app/src/main/java/com/amaze/filemanager/utils/files/GenericCopyUtil.java

```diff
@@ -325,129 +334,94 @@ public void copy(HybridFileParcelable sourceFile, HybridFile targetFile) throws
-
-                    // not enough bytes left in the channel to read, iterate over each byte and store
-                    // in the buffer
-
-                    // reset the counter bytes
-                    count = 0;
-                    for (int i=0; i<buffer.length && inBuffer.hasRemaining(); i++) {
-                        buffer[i] = inBuffer.get();
-                        count++;
-                    }
-                }
-
-                if (count != -1 && !progressHandler.getCancelled()) {
-
-                    bufferedOutputStream.write(buffer, 0, count);
-                    ServiceWatcherUtil.position = inBuffer.position();
-                } else break;
-
-            }
-        } finally {
-            bufferedOutputStream.flush();
-        }
+        doCopy(Channels.newChannel(bufferedInputStream), Channels.newChannel(bufferedOutputStream));
     }
 
     /**
-     * Inner class responsible for getting a {@link ReadableByteChannel} from the input channel
-     * and to watch over the read progress
+     * Calls {@link #doCopy(ReadableByteChannel, WritableByteChannel)}.
+     *
+     * @see Channels#newChannel(OutputStream)
+     * @param inChannel source
+     * @param bufferedOutputStream target
+     * @throws IOException
      */
-    private class CustomReadableByteChannel implements ReadableByteChannel {
-
-        ReadableByteChannel byteChannel;
-
+    @VisibleForTesting
+    void copyFile(@NonNull FileChannel inChannel, @NonNull BufferedOutputStream bufferedOutputStream)
+            throws IOException {
+        doCopy(inChannel, Channels.newChannel(bufferedOutputStream));
+    }
 
-        CustomReadableByteChannel(ReadableByteChannel byteChannel) {
-            this.byteChannel = byteChannel;
+    @VisibleForTesting
+    void doCopy(@NonNull ReadableByteChannel from, @NonNull WritableByteChannel to) throws IOException {
+        ByteBuffer buffer = ByteBuffer.allocateDirect(DEFAULT_TRANSFER_QUANTUM);
+        long count;
+        while ((from.read(buffer) != -1 || buffer.position() > 0) && !progressHandler.getCancelled()) {
+            buffer.flip();
+            count = to.write(buffer);
+            ServiceWatcherUtil.position += count;
+            buffer.compact();
         }
 
-        @Override
-        public int read(ByteBuffer dst) throws IOException {
-            int bytes;
-            if (((bytes = byteChannel.read(dst))>0) && !progressHandler.getCancelled()) {
-
-                ServiceWatcherUtil.position += bytes;
-                return bytes;
+        buffer.flip();
+        while(buffer.hasRemaining())
+            to.write(buffer);
 
-            } else
-                return 0;
-        }
+        from.close();
+        to.close();
+    }
 
-        @Override
-        public boolean isOpen() {
-            return byteChannel.isOpen();
+    @VisibleForTesting
+    void doCopy(@NonNull ReadableByteChannel from, @NonNull FileChannel to) throws IOException{
```

Comment:
Do we really need this method? AFAIK, ````FileChannel```` implements both ````WritableByteChannel```` and ````ReadableByteChannel````.

Reply:
`transferFrom` is implemented in FileChannel.

https://developer.android.com/reference/java/nio/channels/FileChannel#transferFrom(java.nio.channels.ReadableByteChannel,%20long,%20long)

Reply:
IMO it is better to leave as much implementation as possible to Android, especially when dealing with the file system.